keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md

```diff
@@ -0,0 +1,853 @@
+capacity. This has a few consequences:
+
+* Several of the devices might have similar metadata, resulting in a lot of
+  duplication between the published devices.
+* The size of the data required to specify each device reduces the number of
+  devices that can be defined in a single ResourceSlice without hitting the
+  Kubernetes size limit.
+
+The latter can be addressed by splitting the devices across multiple
+`ResourceSlice`s within a single pool, but that isn't always an option.
+In particular, DRA currently doesn't allow sharing counters across `ResourceSlice`s,
+meaning that the number of devices that can fit into a single `ResourceSlice`
+also limits the number of partitionable devices for a single physical device.
+
+### Goals
+
+- Enable a more compact way to define devices in ResourceSlices so duplication can
+  be reduced and a larger number of devices can be published within a single
+  ResourceSlice.
+
+### Non-Goals
+
+- Not part of the plan for alpha: developing kubectl command or plugin to let
+  users see the flattened device definitions. Mixins does make it harder to find
+  the full definition for a specific device, so this might be added to the scope
+  for Beta or GA.
+
+## Proposal
+
+The proposal has two parts to it, the definition of mixins and the
+mechanism for referencing mixins from devices and counter sets.
+
+A new `Mixins` field will be added to the `ResourceSliceSpec` as an
+optional field. It will have three properties, one for each of the three
+types of mixins that will be supported:
+
+1. The `CounterSet` field defines a list of named `CounterSetMixins`.
+   These define counters that can be used to extend the counters
+   explicitly defined in a `CounterSet`. This allows for reduced duplication
+   if there are many identical physical devices that must be represented as
+   `CounterSet`s. `CounterSetMixins` can not be referenced directly by devices.
+
+1. The `Device` field is a list of named `DeviceMixin`s. These define
+   attributes and capacities that can be used to extend what is defined
+   explicitly in `Device`. `DeviceMixin`s cannot be allocated directly, but can
+   only be referenced by devices.
+
+1. The `DeviceCounterConsumption` field defines a list of named
+   `DeviceCounterConsumptionMixin`s. These define counters that can be
+   used to extend the counter consumption defined explicitly `DeviceCounterConsumption`
+   objects in the `ConsumesCounters` list on `Device`. The `CounterSet` from
+   which the counters will be consumed is not specified in the
+   `DeviceCounterConsumptionMixin`, but rather provided when the mixin is
+   referenced from the device.
+
+The mixins are referenced using the same pattern in all three places. The field
+is named `Includes` and will contain a list of references to the mixins. Rather
+than representing the references as a list of strings, there is a list of type
+`<SomeType>MixinRef` which has a `Name` field. This enables adding additional
```

Comment:
So the YAML will be:
```
includes:
- name: something
```

Would it look better to avoid the generic "name" field name and instead use "mixin"?

```
includes:
- mixin: something
```

For someone who doesn't know what "includes" means it might provide an additional hint.

Reply:
The problem with just a `[]string` is that we have to be absolutely sure that we are never going to need anything other than a string for an entry. Using a struct is less risky.

Having said that, I agree that it seems safe here and simplifies the API.

Reply:
I've left it as it is for now. We can bet input from @johnbelamaric and the API reviewer here.

Reply:
I also do not have any ideas on why we would need a struct rather than a string. If we go with struct, I do prefer `mixin: foo` rather than `name: foo`

Reply:
I have updated the KEP use use strings instead of structs.